File: python/src/problems301_400/is_subsequence.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def is_subsequence(self, s: str, t: str) -> bool:
        pass


class Default(Problem):
    @override
    def is_subsequence(self, s: str, t: str) -> bool:
        if not s:
            return True
        if not t:
            return False
        si: int = 0
        ti: int = 0
        while ti < len(t):
            if s[si] == t[ti]:
                si += 1
                if si == len(s):
                    return True
            ti += 1
        return False


SOLUTIONS: list[Problem] = [Default()]

File: typescript/src/problems301_400/is-subsequence.ts
class Problem {
  isSubsequence(s: string, t: string): boolean {
    throw new Error(`Not implemented (${s}, ${t}).`);
  }
}

class Default extends Problem {
  isSubsequence(s: string, t: string): boolean {
    if (s.length === 0) {
      return true;
    }
    if (t.length === 0) {
      return false;
    }
    let si: number = 0;
    let ti: number = 0;
    while (ti < t.length) {
      if (s[si] === t[ti]) {
        si++;
        if (si === s.length) {
          return true;
        }
      }
      ti++;
    }
    return false;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
